File: admin/src/types/floorMap.ts
/**
 * Floor Map v2 类型定义
 * 严格对齐 docs/backend/schemas/floor_map.v2.schema.json
 *
 * 注意：
 *  - StructureType 仅 6 种（column 是独立联合分支，使用 point 而非 rect）
 *  - schema 顶层 additionalProperties:false，不允许写入 floor_map_updated_at（通过 ETag 头部承载）
 *  - PUT 请求时所有 structure 的 source 必须为 'manual'
 */

export type StructureType =
  | 'core'
  | 'elevator'
  | 'stair'
  | 'restroom'
  | 'equipment'
  | 'corridor'

export type AnyStructureType = StructureType | 'column'

export type Source = 'auto' | 'manual'

export interface Rect {
  x: number
  y: number
  w: number
  h: number
}

export interface Structure {
  type: StructureType
  rect: Rect
  label?: string | null
  /** elevator 必填，正则 ^[A-Z]\d{1,3}$ */
  code?: string | null
  /** restroom 必填 */
  gender?: 'M' | 'F' | 'unknown' | null
  source: Source
  /** 仅 source=auto 时存在，[0, 1] */
  confidence?: number
}

export interface Column {
  type: 'column'
  /** 注意：column 用 point，不用 rect */
  point: [number, number]
  source: Source
  confidence?: number
}

export type StructureOrColumn = Structure | Column

export interface WindowSegment {
  side: 'N' | 'S' | 'E' | 'W'
  /** 沿所属边的偏移（像素） */
  offset: number
  /** 沿所属边的长度（像素），≥ 8 */
  width: number
}

/** schema 限制长度 3~32 */
export type PointArray = [number, number][]

export interface Outline {
  type: 'rect' | 'polygon'
  rect?: Rect
  points?: PointArray
}

export interface Unit {
  unit_id: string
  label: string
  polygon: PointArray
  centroid?: [number, number]
  area_sqm?: number
  property_type?: 'office' | 'retail' | 'apartment'
}

export interface FloorMapV2 {
  schema_version: '2.0'
  render_mode: 'vector' | 'semantic'
  viewport: { width: number; height: number }
  outline: Outline
  structures: StructureOrColumn[]
  windows?: WindowSegment[]
  north?: { x: number; y: number; rotation_deg?: number } | null
  units?: Unit[]
  floor_id?: string | null
  building_id?: string | null
  floor_label?: string | null
  svg_version?: string | null
  dxf_region?: { min_x: number; min_y: number; max_x: number; max_y: number } | null
}

/** type guard：判断 StructureOrColumn 是否为 column */
export const isColumn = (s: StructureOrColumn): s is Column => s.type === 'column'

/** PATCH /api/floors/:id/render-mode 响应体 */
export interface RenderModeChangeResult {
  floor_id: string
  render_mode: 'vector' | 'semantic'
  render_mode_changed_at: string
  changed_by: string
}

File: backend/lib/modules/assets/models/floor_map.dart
/// FloorMap 楼层结构数据模型 — 对应 floor_maps 表（migration 027）
///
/// 与 [Floor] 1:1 关联，存储 DXF 抽取的候选结构（candidates）以及
/// 人工审核确认后的结构（structures/outline/windows/north）。
///
/// 关联规范：
/// - docs/backend/FLOOR_MAP_API_SPEC.md
/// - docs/backend/schemas/floor_map.v2.schema.json
class FloorMap {
  final String floorId;
  /// 当前固定 '2.0'，与 floor_map.v2.schema.json 对齐
  final String schemaVersion;
  /// 视口尺寸 { width, height }
  final Map<String, dynamic>? viewport;
  /// 外轮廓 { type: 'rect'|'polygon', rect?, points? }
  final Map<String, dynamic>? outline;
  /// 已审核结构数组（StructureOrColumn 联合类型）
  final List<Map<String, dynamic>> structures;
  /// 窗洞数组 [{ side, offset, width }]
  final List<Map<String, dynamic>> windows;
  /// 指北针 { x, y, rotation_deg? }
  final Map<String, dynamic>? north;
  /// DXF 自动抽取的候选结构（仅 source='auto'），前端只读
  final Map<String, dynamic>? candidates;
  /// 候选项最近一次抽取时间
  final DateTime? candidatesExtractedAt;
  /// 最近一次人工保存时间，作为乐观锁版本号
  final DateTime updatedAt;
  /// 最近一次保存的操作人 user_id
  final String? updatedBy;

  const FloorMap({
    required this.floorId,
    this.schemaVersion = '2.0',
    this.viewport,
    this.outline,
    this.structures = const [],
    this.windows = const [],
    this.north,
    this.candidates,
    this.candidatesExtractedAt,
    required this.updatedAt,
    this.updatedBy,
  });

  factory FloorMap.fromColumnMap(Map<String, dynamic> map) {
    return FloorMap(
      floorId: map['floor_id'] as String,
      schemaVersion: (map['schema_version'] as String?) ?? '2.0',
      viewport: _asMap(map['viewport']),
      outline: _asMap(map['outline']),
      structures: _asListOfMap(map['structures']),
      windows: _asListOfMap(map['windows']),
      north: _asMap(map['north']),
      candidates: _asMap(map['candidates']),
      candidatesExtractedAt: map['candidates_extracted_at'] as DateTime?,
      updatedAt: map['updated_at'] as DateTime,
      updatedBy: map['updated_by'] as String?,
    );
  }

  /// 转换为 API 响应 body（floor_map.v2.schema 兼容）
  ///
  /// 注意：版本号 [updatedAt] 不写入 body，由 Controller 通过 HTTP `ETag`
  /// 响应头返回（schema 顶层 `additionalProperties: false`）。
  Map<String, dynamic> toJson() => {
        'schema_version': schemaVersion,
        'viewport': viewport,
        'outline': outline,
        'structures': structures,
        'windows': windows,
        'north': north,
      };
}

Map<String, dynamic>? _asMap(dynamic v) {
  if (v == null) return null;
  if (v is Map<String, dynamic>) return v;
  if (v is Map) return Map<String, dynamic>.from(v);
  return null;
}

List<Map<String, dynamic>> _asListOfMap(dynamic v) {
  if (v == null) return const [];
  if (v is List) {
    return v
        .whereType<Map>()
        .map((e) => Map<String, dynamic>.from(e))
        .toList(growable: false);
  }
  return const [];
}
